SU2_CFD/src/solvers/CSolver.cpp

```diff
@@ -2065,17 +2068,22 @@ void CSolver::AdaptCFLNumber(CGeometry **geometry,
     /* Check whether we achieved the requested reduction in the linear
      solver residual within the specified number of linear iterations. */
 
-    bool reduceCFL = false;
-    su2double linResFlow = solverFlow->GetResLinSolver();
-    su2double linResTurb = -1.0;
-    if ((iMesh == MESH_0) && (config->GetKind_Turb_Model() != NONE)) {
-      linResTurb = solverTurb->GetResLinSolver();
-    }
+    su2double linResTurb = 0.0;
+    if ((iMesh == MESH_0) && solverTurb) linResTurb = solverTurb->GetResLinSolver();
 
-    su2double maxLinResid = max(linResFlow, linResTurb);
-    if (maxLinResid > 0.5) {
-      reduceCFL = true;
-    }
+    unsigned short linIterTurb = 0;
+    if ((iMesh == MESH_0) && solverTurb) linIterTurb = solverTurb->GetIterLinSolver();
+
+    /* Max residual and iterations between flow and turbulence. */
+    const su2double linRes = max(solverFlow->GetResLinSolver(), linResTurb);
+    const auto linIter = max(solverFlow->GetIterLinSolver(), linIterTurb);
+
+    /* Tolerance limited to a "reasonable" value. */
+    const su2double linTol = max(0.001, config->GetLinear_Solver_Error());
+
+    /* Check if we should decrease or if we can increase, the 20% is to avoid flip-flopping. */
+    const bool reduceCFL = linRes > 1.2*linTol;
+    const bool canIncrease = (linIter <= config->GetLinear_Solver_Iter()) && (linRes < linTol);
```

Comment:
This is my proposal to avoid the saw-tooth problem, here reduce means "apply the reduction factor" rather than returning to the minimum.
If "canIncrease" is not true the increase factor is not applied.

Reply:
This logic makes sense. 

would `linIter <= config->GetLinear_Solver_Iter()` always be true, since the number of linear iterations would be <= the limit specified in the config?

Reply:
That thought popped into my head when I went to sleep last night... It should be < and the `linRes < linTol` is probably redundant as well.

Reply:
Or you could just have  `linRes < linTol`, to take care of the case where it hits the tolerance on the last iteration of the linear solver. I'll test this new adaptive CFL and give this the :+1:

Reply:
Yep I made that change locally. I'm having a look at the criteria to detect stagnation, on the case I am using for testing it is getting triggered but the convergence is fine....
These last changes caused a massive change in residuals for the UQ cases, I have to look into that as well.